### ISSUE TITLE
Proactive entity raw data reduction

### ISSUE THREAD

**Issue by austinkelleher** [JupiterOne/sdk#654]
# Problem

Large entities are currently problematic for our integrations. The JupiterOne backend system has a maximum payload size of 6MB. 6MB is quite large, but we tend to encounter entities that contain a significant amount of raw data and/or large property values. In these cases, the J1 backend system will reject with a `RequestEntityTooLargeException`. When we receive this exception, the SDK [will bubble the exception up](https://github.com/JupiterOne/sdk/blob/1504c547765bd8636a8adcaf19098bc1b9f802e1/packages/integration-sdk-runtime/src/synchronization/index.ts#L278) and the executions step will fail.

Our traditional workaround has been to strip out raw data from the entity type that is causing the large uploads. Unfortunately, the frequency of this issue is quite high and therefore causes frequent integration job failures. The tradeoff of having raw data is not worth the consequence of causing a complete job failure. We should consider changing the upload logic to reduce the entity payload size when uploads fail.

# Proposal - Proactive Entity Raw Data Reduction

When an upload is rejected with `RequestEntityTooLargeException`, the entities are iterated and the records are reduced in size. The request is attempted again with a the reduced data size.

## Example Case

Consider the following example:

- Entity upload delivered]
- JupiterOne backend responds with a 400 `RequestEntityTooLargeException`
- Iterate all entities in upload
- Determine which "top-level" `_rawData` property is largest (see next section on raw data reduction for more details...)
  - Mark the top-level `_rawData` property as "removed." For example: `{ "prop": "REMOVED_TOO_LARGE" }`
- Locally cache the path to the top-level `_rawData` properties that were removed
- Retry the upload
- If the upload fails again, repeat the reduction steps...

**NOTE** - A very similar approach should be easy to apply to relationships as well, but entities are the graph objects that have caused us most (all?) of the issues. Relationships do not have raw data and tend to have very few properties.

### Entity Raw Data Reduction

**Question**: Why not just remove _all_ raw data entirely?
**Answer**: Raw data is a feature of the JupiterOne platform that we've found to be useful. Stripping it out entirely for all entities isn't a great solution when only a subset of entities are problematic. It would be great to keep _most_ of the raw data when possible!

When iterating the "top-level" properties in raw data, we should be mindful that `createIntegrationEntity` writes the `_rawData` property in a specific way by default. It's possible to override this behavior, but most callers are _not_ overriding the behavior. We should consider both cases.

Example Default:

```json
{
  "_rawData": [
    {
      "name": "default":
      "rawData": {
        "firstName": "Austin"
      }
    }
  ]
}
```

In the above case, we'd only want to consider "top-level" entity `_rawData` reduction inside of the `rawData` key.

Example Overridden:

```json
{
  "_rawData": [
    {
      "firstName": "Austin"
    }
  ]
}
```

In the above case, we'd consider "top-level" entity `_rawData` reduction any of the top-level object keys (e.g. `firstName`).

## Future Considerations

It would be great if we could leverage the cached properties to proactively remove these properties from entities when added to the job state, so that the large data doesn't get written to disk. The problem with doing this is that some later steps may rely on the raw data properties that are stripped out.

Additionally, we should consider automatically truncating top-level entity properties.  We'd determine which top-level entity `string` properties are too large (> 4096 characters). Truncate these properties using 
    [`truncateEntityPropertyValue`](https://github.com/JupiterOne/sdk/blob/50867277c9fb06a17b85193dfc0fd7ce992a10bf/packages/integration-sdk-core/src/data/converters.ts#L205). The backend system already truncates these properties, so we could do this a bit more proactively.

**Comment by VDubber** [comment]
> Determine which "top-level" _rawData property is largest

We should also consider that the property also needs to be of significant size. Some entities may not be adding to the problem. It'd be awesome to determine which properties _of which entities_ need to be removed.

> If the upload fails again, repeat the reduction steps...

Can we be more proactive in determining how much data needs to be removed? And calculate how much as been removed prior to reattempting the upload?

> what if we bundled/compressed rawData for customers to download later?

Chris' thought sparked this question: What if we do two uploads? Would it be _crazy_ to always upload all entities without rawData and then update them with the rawData in a separate upload?

**Comment by austinkelleher** [comment]
Fixed in #659
